src/burnchains/db.rs

```diff
@@ -147,16 +262,554 @@ impl<'a> BurnchainDBTransaction<'a> {
+                        "Commit {},{},{} skips reward cycle {} with anchor block",
+                        &block_commit.block_header_hash,
+                        block_commit.block_height,
+                        block_commit.vtxindex,
+                        rc
+                    );
+                    am.push(AffirmationMapEntry::PoxAnchorBlockAbsent);
+                } else {
+                    // affirmation weight increases even if there's no decision made, because
+                    // the lack of a decision is still an affirmation of all prior decisions
+                    test_debug!(
+                        "Commit {},{},{} skips reward cycle {} without anchor block",
+                        &block_commit.block_header_hash,
+                        block_commit.block_height,
+                        block_commit.vtxindex,
+                        rc
+                    );
+                    am.push(AffirmationMapEntry::Nothing);
+                }
+            }
+
+            test_debug!(
+                "Prepare-phase commit {},{},{} affirms parent {},{} with {} descended from {:?}",
+                &block_commit.block_header_hash,
+                block_commit.block_height,
+                block_commit.vtxindex,
+                parent.block_height,
+                parent.vtxindex,
+                &am,
+                &parent_metadata.anchor_block_descendant
+            );
+
+            (am, parent_rc_opt)
+        };
+
+        if let Some(am_id) = BurnchainDB::get_affirmation_map_id(&self.sql_tx, &am)
+            .map_err(|e| BurnchainError::from(e))?
+        {
+            // child doesn't represent any new affirmations by the network, since its
+            // affirmation map already exists.
+            if cfg!(test) {
+                let _am_weight = BurnchainDB::get_affirmation_weight(&self.sql_tx, am_id)?
+                    .expect(&format!("BUG: no affirmation map {}", &am_id));
+
+                test_debug!("Affirmation map of prepare-phase block-commit {},{},{} (parent {},{}) is old: {:?} weight {} affirmed {:?}",
+                            &block_commit.txid, block_commit.block_height, block_commit.vtxindex, block_commit.parent_block_ptr, block_commit.parent_vtxindex, &am, _am_weight, &affirmed_reward_cycle);
+            }
+
+            self.update_block_commit_affirmation(block_commit, affirmed_reward_cycle, am_id)
+                .map_err(|e| BurnchainError::from(e))?;
+            Ok(am_id)
+        } else {
+            test_debug!("Affirmation map of prepare-phase block-commit {},{},{} (parent {},{}) is new: {:?} weight {} affirmed {:?}",
+                        &block_commit.txid, block_commit.block_height, block_commit.vtxindex, block_commit.parent_block_ptr, block_commit.parent_vtxindex, &am, am.weight(), &affirmed_reward_cycle);
+
+            let am_id = self
+                .insert_block_commit_affirmation_map(&am)
+                .map_err(|e| BurnchainError::from(e))?;
```

Comment:
Note you don't need to do a `map_err` call if `From` is implemented. The try operator `?` logically expands to `.map_err(|e| e.into())?`.

Reply:
Fixed